### ISSUE TITLE
f shows hints for a link, but hint completion does open it: a limitation?

### ISSUE THREAD

**Issue by gnarendran** [philc/vimium#2545]
vimium is extremely helpful, thanks a lot. This is a minor issue I notice in some websites, for example in https://chromereleases.googleblog.com/search/label/Stable%20updates

The trouble is with the "comments" links. They open if clicked with mouse, or if chrome's native Search highlighting and Ctrl+Enter is used.

But vimium's f and friends show the hints for the links, give visual indication for the link completion, but the link does not open. I was also able to see vimium was delivering mouseover/mousedown/mouseup/click events for that element. So may be this is a chrome limitation that vimium can't do anything about?

**Comment by gdh1995** [comment]
## Conclusion:

It's a Vimium's bug.

## Details:
Do you mean a link like "2 comments" on this page? Like this?
![image](https://user-images.githubusercontent.com/5547703/27623776-9ec55460-5c10-11e7-8d83-c980ee9dead9.png)

This link is inside an `<iframe>`, then there're some problems if you want to click it:
* There's another hint for this iframe on the top frame, and it covers 2 hints in the small iframe, and you can only see it
  * so Vimium will just "click" the outer iframe, but not focus it. This means nothing useful will be done.
* If you use Chrome Developer Tools, you can find the real 2 hints in the small iframe, but...

Let's assume the top frame is focused when you activate LinkHints, and you type the real hint string
* Vimium will dispatch events about click to the `<a>` element directly, without focusing it, of course
* Chrome does nothing, mainly because the iframe is from a thrid-party website, and has no "focus".
* Nothing useful will be done.

I think this issue shows 2 problems:
* the outer frame should not generate a hint for `<iframe>`, even if it's clickable (e.g. by `tabindex="0"`)
* `activateLink` should dispatch focus to `<a>` after the `mousedown` event, or at least the local `window`.

**Comment by gnarendran** [comment]
Yes indeed, I am referring to links such as '2 comments'. Another thing to note is that vimium's search with / fails to find this string, while native chrome search (Ctrl+F) finds it correctly.

**Comment by gdh1995** [comment]
Vimium can not search across frames, because there's no browser API for this function.

**Comment by gnarendran** [comment]
Oh ok, so / cannot find such strings (which are rare anyway) due to chrome limitation. But f etc., could be fixed to follow such links. Thanks for looking into this and responding!

**Comment by gdh1995** [comment]
> the outer frame should not generate a hint for <iframe>, even if it's clickable (e.g. by tabindex="0")
> activateLink should dispatch focus to <a> after the mousedown event, or at least the local window.

@mrmr1993 What do you think about these ideas? If they're suitable, I can write a PR.

**Comment by mrmr1993** [comment]
>>     the outer frame should not generate a hint for <iframe>, even if it's clickable (e.g. by tabindex="0")
    activateLink should dispatch focus to after the mousedown event, or at least the local window.
>
> @mrmr1993 What do you think about these ideas? If they're suitable, I can write a PR.

I'm not really active in the project any more, so you're probably better off asking @smblott-github or @philc. However, since I *am* highly opinionated:
* frame link hints are useful because they let the user move focus to a scrollable frame
* this frame can't scroll, and so we can't really justify giving it a link hint (even if, as you say, it has eg. `tabindex` set)
* other frames that do scroll still need to have hints, so a PR should take care not to remove them too.

> * Chrome does nothing, mainly because the iframe is from a thrid-party website, and has no "focus".

This is strange, I would still expect our simulated clicks to work when the frame is not focused (so much as to call it a Chromium issue). Does the link hint work when it is?

> * `activateLink` should dispatch focus to `<a>` after the `mousedown` event, or at least the local `window`.

This seems reasonable, if it matches the behaviour of a genuine mouse click.

**Comment by gdh1995** [comment]
> the outer frame should not generate a hint for <iframe>

It's my fault that I didn't explain my real purpose. I think `<iframe>` doesn't need a hint because its `<body>` has provided a hint. But this change has some other issues, like "what if an `<iframe> is not Vimium-enabled".

> Chrome does nothing, mainly because the iframe is from a thrid-party website, and has no "focus".

Sorry again because the "mainly because" is only my inference. In further tests I found:
* After refreshing the whole tab, Chrome will follow the `<a>` on `LinkHints` simulating click events
* However, if the iframe's contentWindow is reloaded (by Vimium's "reload" command or `location.reload()` on Developer Tools), then it does nothing to only simulate mousedown/mouseup/click events without a "focus" event.

There're indeed many bugs in Chrome about the feature rendering third-party iframes in other processes, like broken `contentwindow.postMessage` before Chrome 49 and error positions of the IME bar. So I think this is also one edge case.
